pydantic/main.py

```diff
@@ -313,6 +313,7 @@ def __new__(mcs, name, bases, namespace, **kwargs):  # noqa C901
             '__schema_cache__': {},
             '__json_encoder__': staticmethod(json_encoder),
             '__custom_root_type__': _custom_root_type,
+            '__hash__': (lambda self: hash(tuple(self.__dict__.values()))) if not config.allow_mutation else None,
```

Comment:
Instead of `.values` we should probably have `.items` here.  The keys ought to be considered a part of the identity of the object.

Example:
```
x = {"a": 1, "b": True}
y = {"d": 1, "c": True}
hash(tuple(x.values())) == hash(tuple(y.values()))  # True
hash(tuple(x.items())) == hash(tuple(y.items()))  # False
```

Reply:
In that case, we diverge from behaviour of the built-in dataclass. But I do not know if it is a problem.
If it not a problem, then I agree with you ! :)

Reply:
I don't think this needs to be a lambda, it can be proper method on `BaseModel` I imagine. It could raise an error if `allow_mutation is True` .

Reply:
> I don't think this needs to be a lambda, it can be proper method on `BaseModel` I imagine. It could raise an error if `allow_mutation is True`.

The 'correct' way to make an object unhashable is by setting its `__hash__` attribute to `None`:

```
In [1]: class Foo:
   ...:     pass
   ...:

In [2]: class Bar:
   ...:     __hash__ = None
   ...:

In [3]: hash(Foo())
Out[3]: 284941123

In [4]: hash(Bar())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-d413494abdfb> in <module>
----> 1 hash(Bar())

TypeError: unhashable type: 'Bar'
```

This would have to happen on the metaclass and is done automatically by the Python interpreter for classes which implement `__eq__` but not `__hash__`:

```
In [5]: from pydantic import BaseModel

In [6]: class Baz(BaseModel):
   ...:     pass
   ...:

In [7]: hash(Baz())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-67f7cb492b68> in <module>
----> 1 hash(Baz())

TypeError: unhashable type: 'Baz'

In [8]: Baz.__hash__

In [9]: class Bee:
   ...:     def __eq__(self, other):
   ...:         ...
   ...:

In [10]: Bee.__hash__

In [11]: Foo.__hash__
Out[11]: <slot wrapper '__hash__' of 'object' objects>
```

If `__hash__` were implemented on `BaseModel`, we'd have to emulate the default behaviour.  In addition,  if `__hash__` is not `None`, `isinstance(obj, typing.Hashable)` will not have the desired effect:

```
In [12]: from typing import Hashable

In [13]: isinstance(Bee(), Hashable)
Out[13]: False

In [14]: class FauxUnhashable:
    ...:     def __hash__(self):
    ...:         raise TypeError
    ...:

In [51]: isinstance(FauxUnhashable(), Hashable)
Out[51]: True
```

Reference: https://docs.python.org/3/reference/datamodel.html#object.__hash__

Reply:
okay, so it can't just be a method on `BaseModel`, but let's make it a proper function, not a lambda.

Reply:
> Instead of `.values` we should probably have `.items` here. The keys ought to be considered a part of the identity of the object.

I think this would unnecessary if we include a reference to the model class.